magicctapipe/reco/estimators.py

```diff
@@ -55,76 +53,81 @@ def __init__(
         """
 
         self.telescope_rfs.clear()
 
-        event_data.dropna(subset=self.features, inplace=True)
-
-        if self.use_unsigned_features:
-            event_data[self.features] = np.abs(event_data[self.features])
-
-        # Train RFs per telescope:
-        tel_ids = np.unique(event_data.index.get_level_values("tel_id"))
+        # Train a RF per telescope
+        tel_ids = np.unique(event_data["tel_id"])
 
         for tel_id in tel_ids:
 
-            df_events = event_data.query(f"tel_id == {tel_id}")
+            df_events = event_data.query(f"tel_id == {tel_id}").copy()
+            df_events.dropna(subset=self.features, inplace=True)
+
+            if self.use_unsigned_features:
+                x_train = np.abs(df_events[self.features].to_numpy())
+            else:
+                x_train = df_events[self.features].to_numpy()
 
-            x_train = df_events[self.features].to_numpy()
+            # Use logarithmic energy for the target values
             y_train = np.log10(df_events["true_energy"].to_numpy())
             weights = df_events["event_weight"].to_numpy()
 
             regressor = sklearn.ensemble.RandomForestRegressor(**self.settings)
 
-            logger.info(f"Telescope {tel_id}...")
+            logger.info(f"Training {TEL_NAMES[tel_id]} RF...")
             regressor.fit(x_train, y_train, sample_weight=weights)
 
             self.telescope_rfs[tel_id] = regressor
 
     def predict(self, event_data):
         """
-        Reconstructs the energies of input events with trained RFs.
+        Reconstructs the energies of primary particles with trained RFs.
 
         Parameters
         ----------
         event_data: pandas.core.frame.DataFrame
-            Pandas data frame of shower events
+            Data frame of shower events
 
         Returns
         -------
         reco_params: pandas.core.frame.DataFrame
-            Pandas data frame of reconstructed energies
+            Data frame of the shower events with reconstructed energies
         """
 
-        reco_params = pd.DataFrame()
+        reco_params = pd.DataFrame(
+            data={"reco_energy": [], "reco_energy_var": []},
+            index=pd.MultiIndex.from_tuples([], names=event_data.index.names),
+        )
 
-        event_data.dropna(subset=self.features, inplace=True)
+        # Apply trained RFs to the input data
+        for tel_id, telescope_rf in self.telescope_rfs.items():
 
-        if self.use_unsigned_features:
-            event_data[self.features] = np.abs(event_data[self.features])
+            df_events = event_data.query(f"tel_id == {tel_id}").copy()
+            df_events.dropna(subset=self.features, inplace=True)
 
-        # Apply trained RFs per telescope:
-        tel_ids = np.unique(event_data.index.get_level_values("tel_id"))
+            if df_events.empty:
+                continue
 
-        for tel_id in tel_ids:
+            if self.use_unsigned_features:
```

Comment:
as in the comment above this can be written in a slightly more compact way

Reply:
Thanks, changed.